### PR TITLE
[L1T] [Phase-2] Update Correlator code (Superseding #40670)

### PR DESCRIPTION
**Pull request by aloeliger**
#### PR description:

This PR duplicates and supersedes #40670, as @cecilecaillol has asked me to take it over.

Previous PR description:

> This is an update of the correlator code
> 
> Includes:
> 
> https://github.com/p2l1pfp/cmssw/pull/92 by @thesps and @slaurila
> https://github.com/p2l1pfp/cmssw/pull/94 by @thomreis and @cerminar (only adds a new feature to the emulator, doesn't change any of the existing emulated values)
> https://github.com/p2l1pfp/cmssw/pull/96 by @thesps
> https://github.com/p2l1pfp/cmssw/pull/97 by @drankincms
> https://github.com/p2l1pfp/cmssw/pull/98 by @drankincms (not used by default)
> https://github.com/p2l1pfp/cmssw/pull/100 by @thesps
> https://github.com/p2l1pfp/cmssw/pull/101 by @gpetruc (not used in production)
> https://github.com/p2l1pfp/cmssw/pull/105 by @gpetruc (not used in production)
> https://github.com/p2l1pfp/cmssw/pull/107 by @gpetruc (transparent)
> 
> Local: https://github.com/cms-l1t-offline/cmssw/pull/1063

I will be updating this PR as requested in the original thread, as soon as I am able.

### CODE REVIEW

**Comment by aloeliger** [comment]
I am working on removing the git history of the model and root files.

**Comment by aloeliger** [comment]
Every attempted version of removing the disallowed files (L1Trigger/Phase2L1ParticleFlow/data/modelTT_PUP_Off_dXY_XYCut_Graph.pb & L1Trigger/Phase2L1ParticleFlow/data/jecs/jecs_20220308.root) from the history has either not worked, or left the entirety of CMSSW's history unacceptably rewritten. I'm not sure that I see any real option any more other than squashing the commits to overwrite the history of this branch.

I know Cecile was hesitant to do that. Would the authors of these commits, @thesps @slaurila @thomreis @cerminar @drankincms & @gpetruc  please comment if there is serious objection to that? I'll leave this open for a little while before doing that. If there is objection, I need a method of removing those files from the history, while only changing the respective commit history. Attempts to use methods recommended here: https://www.baeldung.com/git-remove-file-commit-history have not worked.

**Comment by AdrianoDee** [comment]
@aloeliger if one of the (rightful) concern of squashing is giving the right credits to authors, a partial solution would be to:
- drop https://github.com/cms-sw/cmssw/pull/41260/commits/0f1246455214153e05898e9a11a5cbc496a83aa0 in the rebase process (the commit is only adding the `.*pb` file);
- squash everything after https://github.com/cms-sw/cmssw/pull/41260/commits/a564dafd12cb0d22b89ceb520ecd233a391b6a05 (the commit for the ROOT file) in a single commit;
- have this last commit co-authored by those involved in the original one (see the guide [here](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors))

In this way you would still keep some history and would give credits to all the *squashed* authors.

**Comment by aloeliger** [comment]
Thanks, that may be preferable to just squashing the whole thing.

File: L1Trigger/Phase2L1ParticleFlow/interface/BJetId.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+         const std::string &iWeightFile,
+         int iNParticles);
+  ~BJetId();
+
+  void setNNVectorVar();
+  float EvaluateNN();
+  float compute(const l1t::PFJet &iJet, float vz, bool useRawPt);
+
+private:
+  tensorflow::Session *session_;
```
Comment (srimanob) [inline L1Trigger/Phase2L1ParticleFlow/interface/BJetId.h:28]: Comment from previous PR from @makortel 
https://github.com/cms-sw/cmssw/pull/40670#discussion_r1120877734

**Comment by aloeliger** [comment]
@srimanob  I have not opened any PR for data externals. I've never actually opened a data PR. Where should that be opened?

**Comment by aloeliger** [comment]
> @aloeliger if one of the (rightful) concern of squashing is giving the right credits to authors, a partial solution could be to:
> 
> * drop [0f12464](https://github.com/cms-sw/cmssw/commit/0f1246455214153e05898e9a11a5cbc496a83aa0) in the rebase process (the commit is only adding the `.*pb` file);
> * squash everything after [a564daf](https://github.com/cms-sw/cmssw/commit/a564dafd12cb0d22b89ceb520ecd233a391b6a05) (the commit for the ROOT file) in a single commit;
> * have this last commit co-authored by those involved in the original ones (see the guide [here](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors))
> 
> In this way you would still keep most of the history and would give credits to all the _squashed_ authors.

@thesps @slaurila @thomreis @cerminar @drankincms & @gpetruc I believe i have prepared a version of the branch doing this (Giovanni and Emyr you should be listed as co-authors on squashed commits). Unless I hear from you by tomorrow morning Geneva time, I will for push that version of the branch and work on other comments for this PR.

**Comment by thesps** [comment]
@aloeliger I'm fine with the squash, thanks for preparing it

**Comment by aloeliger** [comment]
@thesps @slaurila @thomreis @cerminar @drankincms & @gpetruc Would it be possible for you to provide the root and model file removed from these PRs? Do these need to be in the data externals? Please let me know so I can open a corresponding PR to provide this data.

**Comment by thesps** [comment]
You can fetch them from https://github.com/p2l1pfp/cmssw/tree/L1PF_12_5_X/L1Trigger/Phase2L1ParticleFlow/data 
(Safer than if I just send you some files)

**Comment by aloeliger** [comment]
@smuzaffar would it take anything special to get L1Trigger/Phase2L1ParticleFlow/data/modelTT_PUP_Off_dXY_XYCut_Graph.pb & L1Trigger/Phase2L1ParticleFlow/data/jecs/jecs_20220308.root integrated into https://github.com/cms-data? These would be the first of the data-files for L1Trigger-Phase2L1ParticleFlow. Is there anything special about creation of the repository or it's name?

**Comment by smuzaffar** [comment]
@aloeliger , nothing special, I just need to create the new data repo. if L1Trigger/Phase2L1ParticleFlow is the right package from these data files then I can create data-L1Trigger-Phase2L1ParticleFlow

**Comment by gpetruc** [comment]
Isn't it already there? https://github.com/cms-data/L1Trigger-Phase2L1ParticleFlow

**Comment by smuzaffar** [comment]
> Isn't it already there? https://github.com/cms-data/L1Trigger-Phase2L1ParticleFlow

oops, you are right :-) @aloeliger repo is already there, you just need to open a PR with new data files there
thanks @gpetruc

**Comment by aloeliger** [comment]
Dang. That messed up the history. Okay. Let me retry this.

**Comment by gpetruc** [comment]
Ok, no problem.

Just put in the commit log and new PR a link to the old PR to
cms-l1t-offline to help trace back history if needed for debugging later.

Il Mer 5 Apr 2023, 09:13 Andrew Loeliger ***@***.***> ha
scritto:

**Pull request closed without merging**